File: packages/typescript/src/envelope.ts
import { randomUUID } from "node:crypto";
import { EventType, isValidEventType, type EventTypeValue } from "./events.js";

const REQUIRED_FIELDS = [
  "openhook",
  "id",
  "source",
  "type",
  "time",
  "session_id",
] as const;

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

export interface OpenHookEventInit {
  source: string;
  type: EventTypeValue;
  sessionId: string;
  data?: Record<string, unknown>;
  cwd?: string;
  extensions?: Record<string, unknown>;
  id?: string;
  time?: string;
}

export class OpenHookEvent {
  readonly openhook: string;
  readonly id: string;
  readonly source: string;
  readonly type: EventTypeValue;
  readonly time: string;
  readonly sessionId: string;
  readonly data: Record<string, unknown>;
  readonly cwd: string | undefined;
  readonly extensions: Record<string, unknown>;

  constructor(init: OpenHookEventInit) {
    this.openhook = "0.1";
    this.id = init.id ?? randomUUID();
    this.source = init.source;
    this.type = init.type;
    this.time = init.time ?? new Date().toISOString();
    this.sessionId = init.sessionId;
    this.data = init.data ?? {};
    this.cwd = init.cwd;
    this.extensions = init.extensions ?? {};
  }

  get transcriptPath(): string | undefined {
    const p = this.data["transcript_path"];
    return typeof p === "string" ? p : undefined;
  }

  get isTrace(): boolean {
    return this.transcriptPath !== undefined;
  }

  get isMetric(): boolean {
    return (
      this.type === EventType.PromptSubmit ||
      this.type === EventType.ToolStart ||
      this.type === EventType.ToolEnd ||
      this.type === EventType.SessionEnd
    );
  }

  static fromObject(d: Record<string, unknown>): OpenHookEvent {
    validate(d);
    return new OpenHookEvent({
      source: d["source"] as string,
      type: d["type"] as EventTypeValue,
      sessionId: d["session_id"] as string,
      data: (d["data"] as Record<string, unknown>) ?? {},
      cwd: d["cwd"] as string | undefined,
      extensions: (d["extensions"] as Record<string, unknown>) ?? {},
      id: d["id"] as string,
      time: d["time"] as string,
    });
  }

  static fromJSON(raw: string): OpenHookEvent {
    return OpenHookEvent.fromObject(JSON.parse(raw));
  }

  toObject(): Record<string, unknown> {
    const obj: Record<string, unknown> = {
      openhook: this.openhook,
      id: this.id,
      source: this.source,
      type: this.type,
      time: this.time,
      session_id: this.sessionId,
    };
    if (Object.keys(this.data).length > 0) obj["data"] = this.data;
    if (this.cwd) obj["cwd"] = this.cwd;
    if (Object.keys(this.extensions).length > 0)
      obj["extensions"] = this.extensions;
    return obj;
  }

  toJSON(): string {
    return JSON.stringify(this.toObject());
  }

  emit(stream: NodeJS.WritableStream = process.stdout): void {
    stream.write(this.toJSON() + "\n");
  }
}

export function validate(d: Record<string, unknown>): void {
  const missing = REQUIRED_FIELDS.filter((f) => !(f in d));
  if (missing.length > 0) {
    throw new ValidationError(
      `Missing required fields: ${missing.sort().join(", ")}`
    );
  }

  if (typeof d["openhook"] !== "string") {
    throw new ValidationError("'openhook' must be a string");
  }

  const typeVal = d["type"];
  if (typeof typeVal !== "string" || !isValidEventType(typeVal)) {
    throw new ValidationError(`Unknown event type: ${JSON.stringify(typeVal)}`);
  }
}

export async function parseStdin(): Promise<OpenHookEvent> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer);
  }
  const raw = Buffer.concat(chunks).toString("utf-8").trim();
  if (!raw) {
    throw new ValidationError("Empty stdin");
  }
  return OpenHookEvent.fromJSON(raw);
}

File: packages/python/src/openhook/envelope.py
"""OpenHook envelope: parse, validate, create, emit."""

from __future__ import annotations

import json
import sys
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .events import EventType

REQUIRED_FIELDS = frozenset({"openhook", "id", "source", "type", "time", "session_id"})


class ValidationError(Exception):
    pass


@dataclass(frozen=True)
class OpenHookEvent:
    openhook: str
    id: str
    source: str
    type: EventType
    time: str
    session_id: str
    data: dict[str, Any] = field(default_factory=dict)
    context: str | None = None
    extensions: dict[str, Any] = field(default_factory=dict)

    # --- Convenience accessors ---

    @property
    def transcript_path(self) -> Path | None:
        p = self.data.get("transcript_path")
        return Path(p) if p else None

    @property
    def is_trace(self) -> bool:
        return self.transcript_path is not None

    @property
    def is_metric(self) -> bool:
        return self.type in (
            EventType.PROMPT_SUBMIT,
            EventType.TOOL_START,
            EventType.TOOL_END,
            EventType.SESSION_END,
        )

    # --- Constructors ---

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> OpenHookEvent:
        validate(d)
        return cls(
            openhook=d["openhook"],
            id=d["id"],
            source=d["source"],
            type=EventType(d["type"]),
            time=d["time"],
            session_id=d["session_id"],
            data=d.get("data", {}),
            context=d.get("context"),
            extensions=d.get("extensions", {}),
        )

    @classmethod
    def from_json(cls, raw: str) -> OpenHookEvent:
        return cls.from_dict(json.loads(raw))

    @classmethod
    def create(
        cls,
        *,
        source: str,
        type: EventType,
        session_id: str,
        data: dict[str, Any] | None = None,
        context: str | None = None,
        extensions: dict[str, Any] | None = None,
        event_id: str | None = None,
        time: str | None = None,
    ) -> OpenHookEvent:
        return cls(
            openhook="0.1",
            id=event_id or str(uuid.uuid4()),
            source=source,
            type=type,
            time=time or datetime.now(timezone.utc).isoformat(),
            session_id=session_id,
            data=data or {},
            context=context,
            extensions=extensions or {},
        )

    # --- Serialization ---

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "openhook": self.openhook,
            "id": self.id,
            "source": self.source,
            "type": str(self.type),
            "time": self.time,
            "session_id": self.session_id,
        }
        if self.data:
            d["data"] = self.data
        if self.context:
            d["context"] = self.context
        if self.extensions:
            d["extensions"] = self.extensions
        return d

    def to_json(self) -> str:
        return json.dumps(self.to_dict())

    def emit(self, file: Any = None) -> None:
        out = file or sys.stdout
        out.write(self.to_json())
        out.write("\n")
        out.flush()


def validate(d: dict[str, Any]) -> None:
    missing = REQUIRED_FIELDS - d.keys()
    if missing:
        raise ValidationError(f"Missing required fields: {', '.join(sorted(missing))}")

    if not isinstance(d.get("openhook"), str):
        raise ValidationError("'openhook' must be a string")

    type_val = d.get("type", "")
    try:
        EventType(type_val)
    except ValueError:
        raise ValidationError(f"Unknown event type: {type_val!r}") from None


def parse_stdin() -> OpenHookEvent:
    raw = sys.stdin.read()
    if not raw.strip():
        raise ValidationError("Empty stdin")
    return OpenHookEvent.from_json(raw)
